File: jacodb-ets/src/main/kotlin/org/jacodb/ets/dto/Values.kt
/*
 *  Copyright 2022 UnitTestBot contributors (utbot.org)
 * <p>
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 * <p>
 *  http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package org.jacodb.ets.dto

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonClassDiscriminator
import kotlinx.serialization.json.JsonObject

@Serializable
@OptIn(ExperimentalSerializationApi::class)
@JsonClassDiscriminator("_")
sealed interface ValueDto {
    val type: TypeDto
}

@Serializable(with = RawValueSerializer::class)
@SerialName("RawValue")
@OptIn(ExperimentalSerializationApi::class)
data class RawValueDto(
    val kind: String, // constructor name
    val extra: JsonObject,
    override val type: TypeDto,
) : ValueDto

@Serializable
sealed interface ImmediateDto : ValueDto

@Serializable
@SerialName("Local")
data class LocalDto(
    val name: String,
    override val type: TypeDto,
) : ImmediateDto

@Serializable
@SerialName("Constant")
data class ConstantDto(
    val value: String,
    override val type: TypeDto,
) : ImmediateDto

// TODO: uncomment and adapt the following code for different constants
//       Currently, ArkIR stores all constants as strings (see ConstantDto above),
//       but some day this is going to be fixed.
//       I expect that the solution on the arkanalyzer's side to be similar to the one below,
//       so I'm keeping it here for future reference.
//
// @Serializable
// sealed interface ConstantDto: ImmediateDto
//
// @Serializable
// @SerialName("StringConstant")
// data class StringConstantDto(
//     val value: String,
// ) : ConstantDto {
//     override val type: Type
//         get() = StringType
//
//     override fun toString(): String = "\"$value\""
// }
//
// @Serializable
// @SerialName("BooleanConstant")
// data class BooleanConstantDto(
//     val value: Boolean,
// ) : ConstantDto {
//     override val type: Type
//         get() = BooleanType
//
//     override fun toString(): String = value.toString()
//
//     companion object {
//         val TRUE = BooleanConstant(true)
//         val FALSE = BooleanConstant(false)
//     }
// }
//
// @Serializable
// @SerialName("NumberConstant")
// data class NumberConstantDto(
//     val value: Double,
// ) : ConstantDto {
//     override val type: Type
//         get() = NumberType
//
//     override fun toString(): String = value.toString()
// }
//
// @Serializable
// @SerialName("NullConstant")
// object NullConstantDto : ConstantDto {
//     override val type: Type
//         get() = NullType
//
//     override fun toString(): String = "null"
// }
//
// @Serializable
// @SerialName("UndefinedConstant")
// object UndefinedConstantDto : ConstantDto {
//     override val type: Type
//         get() = UndefinedType
//
//     override fun toString(): String = "undefined"
// }

@Serializable
sealed interface ExprDto : ValueDto

@Serializable
@SerialName("NewExpr")
data class NewExprDto(
    val classType: TypeDto, // ClassType
) : ExprDto {
    override val type: TypeDto
        get() = classType
}

@Serializable
@SerialName("NewArrayExpr")
data class NewArrayExprDto(
    val elementType: TypeDto,
    val size: ValueDto,
) : ExprDto {
    override val type: TypeDto
        get() = ArrayTypeDto(elementType, 1)
}

@Serializable
@SerialName("DeleteExpr")
data class DeleteExprDto(
    val arg: ValueDto,
) : ExprDto {
    override val type: TypeDto
        get() = BooleanTypeDto
}

@Serializable
@SerialName("AwaitExpr")
data class AwaitExprDto(
    val arg: ValueDto,
) : ExprDto {
    override val type: TypeDto
        get() = arg.type
}

@Serializable
@SerialName("YieldExpr")
data class YieldExprDto(
    val arg: ValueDto,
) : ExprDto {
    override val type: TypeDto
        get() = arg.type
}

@Serializable
@SerialName("TypeOfExpr")
data class TypeOfExprDto(
    val arg: ValueDto,
) : ExprDto {
    override val type: TypeDto
        get() = StringTypeDto
}

@Serializable
@SerialName("InstanceOfExpr")
data class InstanceOfExprDto(
    val arg: ValueDto,
    val checkType: TypeDto,
) : ExprDto {
    override val type: TypeDto
        get() = BooleanTypeDto
}

@Serializable
@SerialName("CastExpr")
data class CastExprDto(
    val arg: ValueDto,
    override val type: TypeDto,
) : ExprDto

@Serializable
sealed interface UnaryExprDto : ExprDto {
    val arg: ValueDto

    override val type: TypeDto
        get() = arg.type
}

@Serializable
@SerialName("UnopExpr")
data class UnaryOperationDto(
    val op: String,
    override val arg: ValueDto,
) : UnaryExprDto

@Serializable
sealed interface BinaryExprDto : ExprDto {
    val left: ValueDto
    val right: ValueDto
}

@Serializable
@SerialName("BinopExpr")
data class BinaryOperationDto(
    val op: String,
    override val left: ValueDto,
    override val right: ValueDto,
    override val type: TypeDto = UnknownTypeDto,
) : BinaryExprDto

@Serializable
sealed interface ConditionExprDto : BinaryExprDto

@Serializable
@SerialName("ConditionExpr")
data class RelationOperationDto(
    val op: String,
    override val left: ValueDto,
    override val right: ValueDto,
    override val type: TypeDto = UnknownTypeDto,
) : ConditionExprDto

@Serializable
sealed interface CallExprDto : ExprDto {
    val method: MethodSignatureDto
    val args: List<ValueDto>

    override val type: TypeDto
        get() = method.returnType
}

@Serializable
@SerialName("InstanceCallExpr")
data class InstanceCallExprDto(
    val instance: ValueDto, // Local
    override val method: MethodSignatureDto,
    override val args: List<ValueDto>,
) : CallExprDto

@Serializable
@SerialName("StaticCallExpr")
data class StaticCallExprDto(
    override val method: MethodSignatureDto,
    override val args: List<ValueDto>,
) : CallExprDto

@Serializable
@SerialName("PtrCallExpr")
data class PtrCallExprDto(
    val ptr: ValueDto, // Local or FieldRef
    override val method: MethodSignatureDto,
    override val args: List<ValueDto>,
) : CallExprDto

@Serializable
sealed interface RefDto : ValueDto

@Serializable
@SerialName("ThisRef")
data class ThisRefDto(
    override val type: TypeDto, // ClassType
) : RefDto

@Serializable
@SerialName("ParameterRef")
data class ParameterRefDto(
    val index: Int,
    override val type: TypeDto,
) : RefDto

// @Serializable
// @SerialName("CaughtExceptionRef")
// data class CaughtExceptionRefDto(
//     override val type: TypeDto,
// ) : RefDto
//
// @Serializable
// @SerialName("GlobalRef")
// data class GlobalRefDto(
//     val name: String,
//     val ref: ValueDto?,
// ) : RefDto {
//     override val type: TypeDto
//         get() = ref?.type ?: UnknownTypeDto
// }
//
// @Serializable
// @SerialName("ClosureFieldRef")
// data class ClosureFieldRefDto(
//     val base: LocalDto,
//     val fieldName: String,
//     override val type: TypeDto,
// ) : RefDto

@Serializable
@SerialName("ArrayRef")
data class ArrayRefDto(
    val array: ValueDto,
    val index: ValueDto,
    override val type: TypeDto,
) : RefDto

@Serializable
sealed interface FieldRefDto : RefDto {
    val field: FieldSignatureDto

    override val type: TypeDto
        get() = this.field.type
}

@Serializable
@SerialName("InstanceFieldRef")
data class InstanceFieldRefDto(
    val instance: ValueDto, // Local
    override val field: FieldSignatureDto,
) : FieldRefDto

@Serializable
@SerialName("StaticFieldRef")
data class StaticFieldRefDto(
    override val field: FieldSignatureDto,
) : FieldRefDto

File: jacodb-ets/ts-frontend/src/dto/values.ts
/*
 *  Copyright 2022 UnitTestBot contributors (utbot.org)
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

/**
 * Value DTOs. Mirror `org.jacodb.ets.dto` values
 * (jacodb-ets/src/main/kotlin/org/jacodb/ets/dto/Values.kt).
 *
 * Serialized polymorphically with the "_" discriminator.
 * Unknown discriminators fall back to RawValueDto on the Kotlin side
 * (RawValue REQUIRES a "type" key).
 *
 * IMPORTANT: only fields that are Kotlin constructor parameters are serialized.
 * Computed `type` getters (e.g. on NewExpr, UnopExpr, call exprs) must NOT be emitted —
 * the strict Kotlin Json would reject the extra key.
 */

import { BinaryOp, RelationOp, UnaryOp } from "./ops";
import { FieldSignatureDto, MethodSignatureDto } from "./signatures";
import { TypeDto } from "./types";

/**
 * Closed union of the known value kinds.
 * Raw fallback values (RawValueDto) are intentionally NOT part of this union:
 * they carry an index signature that would break discriminated-union narrowing.
 * Emission sites must cast them explicitly (see lowering/fallback.ts).
 */
export type ValueDto = ImmediateDto | ExprDto | RefDto;

export type ImmediateDto = LocalDto | ConstantDto;

export type ExprDto =
    | NewExprDto
    | NewArrayExprDto
    | DeleteExprDto
    | AwaitExprDto
    | YieldExprDto
    | TypeOfExprDto
    | InstanceOfExprDto
    | CastExprDto
    | UnopExprDto
    | BinopExprDto
    | ConditionExprDto
    | InstanceCallExprDto
    | StaticCallExprDto
    | PtrCallExprDto;

export type CallExprDto = InstanceCallExprDto | StaticCallExprDto | PtrCallExprDto;

export type RefDto =
    | ThisRefDto
    | ParameterRefDto
    | CaughtExceptionRefDto
    | GlobalRefDto
    | ClosureFieldRefDto
    | ArrayRefDto
    | InstanceFieldRefDto
    | StaticFieldRefDto;

/** LValue kinds accepted by Kotlin Convert as the LHS of AssignStmt. */
export type LValueDto = LocalDto | ArrayRefDto | InstanceFieldRefDto | StaticFieldRefDto;

/** Fallback for constructs we cannot model; Kotlin deserializes any unknown kind into RawValueDto. */
export interface RawValueDto {
    readonly _: string; // any kind not listed above
    type: TypeDto; // REQUIRED by RawValueSerializer
    [extra: string]: unknown;
}

export interface LocalDto {
    readonly _: "Local";
    name: string;
    type: TypeDto;
}

/** All constants carry a string-encoded value (e.g. "42", "true", "null"). */
export interface ConstantDto {
    readonly _: "Constant";
    value: string;
    type: TypeDto;
}

export interface NewExprDto {
    readonly _: "NewExpr";
    classType: TypeDto; // ClassType
}

export interface NewArrayExprDto {
    readonly _: "NewArrayExpr";
    elementType: TypeDto;
    size: ValueDto;
}

export interface DeleteExprDto {
    readonly _: "DeleteExpr";
    arg: ValueDto;
}

export interface AwaitExprDto {
    readonly _: "AwaitExpr";
    arg: ValueDto;
}

export interface YieldExprDto {
    readonly _: "YieldExpr";
    arg: ValueDto;
}

export interface TypeOfExprDto {
    readonly _: "TypeOfExpr";
    arg: ValueDto;
}

export interface InstanceOfExprDto {
    readonly _: "InstanceOfExpr";
    arg: ValueDto;
    checkType: TypeDto;
}

export interface CastExprDto {
    readonly _: "CastExpr";
    arg: ValueDto;
    type: TypeDto;
}

export interface UnopExprDto {
    readonly _: "UnopExpr";
    op: UnaryOp;
    arg: ValueDto;
}

export interface BinopExprDto {
    readonly _: "BinopExpr";
    op: BinaryOp;
    left: ValueDto;
    right: ValueDto;
    type?: TypeDto; // Kotlin default: UnknownType
}

export interface ConditionExprDto {
    readonly _: "ConditionExpr";
    op: RelationOp;
    left: ValueDto;
    right: ValueDto;
    type?: TypeDto; // Kotlin default: UnknownType
}

export interface InstanceCallExprDto {
    readonly _: "InstanceCallExpr";
    instance: LocalDto; // Kotlin Convert casts this to LocalDto — MUST be a Local
    method: MethodSignatureDto;
    args: ValueDto[];
}

export interface StaticCallExprDto {
    readonly _: "StaticCallExpr";
    method: MethodSignatureDto;
    args: ValueDto[];
}

export interface PtrCallExprDto {
    readonly _: "PtrCallExpr";
    ptr: ValueDto; // Local or FieldRef (must be a value, not an expr)
    method: MethodSignatureDto;
    args: ValueDto[];
}

export interface ThisRefDto {
    readonly _: "ThisRef";
    type: TypeDto; // ClassType
}

export interface ParameterRefDto {
    readonly _: "ParameterRef";
    index: number;
    type: TypeDto;
}

export interface CaughtExceptionRefDto {
    readonly _: "CaughtExceptionRef";
    type: TypeDto;
}

export interface GlobalRefDto {
    readonly _: "GlobalRef";
    name: string;
    ref: ValueDto | null; // nullable WITHOUT default on the Kotlin side — must be present
}

export interface ClosureFieldRefDto {
    readonly _: "ClosureFieldRef";
    base: LocalDto;
    fieldName: string;
    type: TypeDto;
}

export interface ArrayRefDto {
    readonly _: "ArrayRef";
    array: ValueDto;
    index: ValueDto; // must be a value (immediate/ref), not an expr — Kotlin casts to EtsValue
    type: TypeDto;
}

export interface InstanceFieldRefDto {
    readonly _: "InstanceFieldRef";
    instance: LocalDto; // Kotlin Convert casts this to LocalDto — MUST be a Local
    field: FieldSignatureDto;
}

export interface StaticFieldRefDto {
    readonly _: "StaticFieldRef";
    field: FieldSignatureDto;
}
